### DISCUSSION TITLE
Denoising & Filling problem

### DISCUSSION THREAD

**Discussion by Antonio-Silveira** [mtex-toolbox/mtex#1395] [Ask Anything] [answered]
Hello everyone,

I am having a hard time denoising my data. Based on the documentation, I tried to remove some very small grains, then smoothed with Halfquadratic filter or Mean filter together with the filling command. 

The problem is, that the grains do not seem to grow over the deleted points.

 I don't understand if this is more related to the data quality, which is not the best, or it's more related to me being a noob. I am attaching the figures from the initial data, Quality map, after I remove small grains and after the Half Quadratic and Mean Filter.

Thanks in advance for any tip :)
Antonio

**Here is my code:**
```matlab
%% Ploting phase & grains
figure(1)
plot(ebsdz)
hold on
[grainsz,ebsdz.grainId] = calcGrains(ebsdz('indexed'),'unitCell','angle',15*degree);
plot(grainsz.boundary,'linewidth',.8)
hold off

%% Quality map
figure(2)
plot(ebsdz,ebsdz.iq)
hold on
mtexColorbar
hold off

%% removing small grains
[grainsz,ebsdz.grainId] = calcGrains(ebsdz('indexed'),'unitCell','angle',15*degree);
ebsdz(grainsz(grainsz.grainSize<3)) = [];
[grainsz,ebsdz.grainId] = calcGrains(ebsdz('indexed'),'unitCell','angle',15*degree);


%% Filtering Halfquadratic
F = halfQuadraticFilter;
F.alpha = 50; %i trying using 0.01, 5, 10 or 50, but nothing really different happened

ebsdzS = smooth(ebsdz('indexed'),F,'fill',grainsz);
[grainszS,ebsdzS.grainId] = calcGrains(ebsdzS('indexed'),'unitCell','angle',15*degree);

figure(3)
plot(ebsdzS)
hold on
plot(grainszS.boundary,'linewidth',.8)
hold off

%% filtering zoomed - MEDIAN FILTER
F = medianFilter;
F.numNeighbours = 3; %tried increasing this value, but the process gets insanely slow


ebsdzSm = smooth(ebsdz('indexed'),F,'fill',grainsz);
[grainszSm,ebsdzSm.grainId] = calcGrains(ebsdzSm('indexed'),'unitCell','angle',15*degree);

figure(4)
plot(ebsdzSm)


hold on
plot(grainszSm.boundary,'linewidth',.8)
hold off
```
**Initial data**

![initial_data_pha_grain](https://user-images.githubusercontent.com/103099688/161924888-251dddb9-618f-49d5-8670-dce36e499307.png)

**Quality Map**
![quality_map](https://user-images.githubusercontent.com/103099688/161924891-5fda927b-645b-4e90-9d5a-4e5440606cdc.png)

**After small grain removal**

![without_small_grains](https://user-images.githubusercontent.com/103099688/161924971-3be8a3e2-403c-4dfd-86f6-f63655fb49bf.png)

**Half Quadratic filter**

![halfquadraticF](https://user-images.githubusercontent.com/103099688/161925021-1ec43139-94a3-4f05-9215-103f30901cb4.png)

**Mean Filter**

![meanF](https://user-images.githubusercontent.com/103099688/161925064-3e65d3d1-9214-42d7-b937-a9ec9b76dd44.png)

**Comment by kilir** [comment, marked as answer]
Hi,
you are telling mtex to fill areas defined by `grainsz`. Since you compute grains using `'unitcell'` the areas you'd try to fill are actualyl occupied by notIndexed grains, preventing them to be filled, also you only direct mtex to smooth the `'indexed'` part. So either just deleting the small areas you like to fill and smoothing on the entire set should help. So I guess you'd want achieve something like case 5 here? https://mtex-toolbox.github.io/GrainReconstructionAdvanced.html

Cheers,
Rüdiger

**Comment by Antonio-Silveira** [comment]
Hi Rüdiger,
I was using ```'unitcell'``` without fully understanding it. After I remove it, it worked.

Thank you Rüdiger :)
